aws/rust-runtime/aws-inlineable/src/http_request_checksum.rs

```diff
@@ -99,40 +103,48 @@ impl DefaultRequestChecksumOverride {
     }
 }
 
-pub(crate) struct RequestChecksumInterceptor<AP> {
+pub(crate) struct RequestChecksumInterceptor<AP, CM> {
     algorithm_provider: AP,
+    checksum_mutator: CM,
 }
 
-impl<AP> fmt::Debug for RequestChecksumInterceptor<AP> {
+impl<AP, CM> fmt::Debug for RequestChecksumInterceptor<AP, CM> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("RequestChecksumInterceptor").finish()
     }
 }
 
-impl<AP> RequestChecksumInterceptor<AP> {
-    pub(crate) fn new(algorithm_provider: AP) -> Self {
-        Self { algorithm_provider }
+impl<AP, CM> RequestChecksumInterceptor<AP, CM> {
+    pub(crate) fn new(algorithm_provider: AP, checksum_mutator: CM) -> Self {
+        Self {
+            algorithm_provider,
+            checksum_mutator,
+        }
     }
 }
 
-impl<AP> Intercept for RequestChecksumInterceptor<AP>
+impl<AP, CM> Intercept for RequestChecksumInterceptor<AP, CM>
 where
-    AP: Fn(&Input) -> Result<Option<ChecksumAlgorithm>, BoxError> + Send + Sync,
+    AP: Fn(&Input) -> Result<(Option<ChecksumAlgorithm>, bool), BoxError> + Send + Sync,
+    CM: Fn(&mut Input, &ConfigBag) -> Result<(), BoxError> + Send + Sync,
 {
     fn name(&self) -> &'static str {
         "RequestChecksumInterceptor"
     }
 
-    fn read_before_serialization(
+    fn modify_before_serialization(
         &self,
-        context: &BeforeSerializationInterceptorContextRef<'_>,
+        context: &mut BeforeSerializationInterceptorContextMut<'_>,
         _runtime_components: &RuntimeComponents,
         cfg: &mut ConfigBag,
     ) -> Result<(), BoxError> {
+        let _ = (self.checksum_mutator)(context.input_mut(), cfg);
         let checksum_algorithm = (self.algorithm_provider)(context.input())?;
-
         let mut layer = Layer::new("RequestChecksumInterceptor");
-        layer.store_put(RequestChecksumInterceptorState { checksum_algorithm });
+        layer.store_put(RequestChecksumInterceptorState {
+            checksum_algorithm: checksum_algorithm.0,
+            request_checksum_required: checksum_algorithm.1,
+        });
```

Comment:
prefer destructuring and concise struct field assignment:
```rust
        let _ = (self.checksum_mutator)(context.input_mut(), cfg);
        let (checksum_algorithm, request_checksum_required) = (self.algorithm_provider)(context.input())?;

        let mut layer = Layer::new("RequestChecksumInterceptor");

        layer.store_put(RequestChecksumInterceptorState {
            checksum_algorithm,
            request_checksum_required,
        });
```

Reply:
Updated in https://github.com/smithy-lang/smithy-rs/commit/b80f97148c9d943509d1a8936ce5637da5ab525e